feat(storage): add RocksDB consistency check on startup

yongkangc opened the pull request:
Addresses [#20392](https://github.com/paradigmxyz/reth/issues/20392) - Initialize RocksDB provider and add consistency check on startup.

## Changes

- Add RocksDB `check_consistency()` call in node launch with correct ordering:
  1. `StaticFileProvider::check_file_consistency()` - heals interrupted writes
  2. `RocksDB::check_consistency()` - uses static file data for tx hash pruning  
  3. `StaticFileProvider::check_consistency()` - compares with MDBX checkpoints


Note for future:

- We have to follow up with passing RocksDB to the stages once implemented in https://github.com/paradigmxyz/reth/issues/20389
- And also we can spin off another pr for Configuration validation

yongkangc committed:
feat(storage): add RocksDB consistency check on startup

Amp-Thread-ID: https://ampcode.com/threads/T-019b2fb1-ce5c-7251-b454-0d7472a0754a

yongkangc committed:
chore: format code

yongkangc committed:
fix: add const to stub check_consistency fn

yongkangc committed:
fix(storage): address review findings for RocksDB consistency check

- Add missing methods to rocksdb_stub.rs (batch, first, last, iter, commit, RocksDBIter)
- Add checkpoint > 0 guard in (None, Some(highest_tx)) case for TransactionHashNumbers
- Use checked_add for mdbx_tx + 1 to prevent u64 overflow
- Update misleading docstring in stub's RocksDBProvider::new

Amp-Thread-ID: https://ampcode.com/threads/T-019b3076-f7ed-700d-8916-779712facb8e

fgimenez commented on crates/node/builder/src/launch/common.rs:
```diff
@@ -501,20 +502,50 @@ where
+                (None, Some(_)) => "RocksDB <> database",
+                (None, None) => unreachable!(),
+            };
             assert_ne!(
```
this will panic when rocksdb returns `Some(0)` as unwind target, which can happen for instance when an existing node enables rocksdb flags. is this what we want? this will be a common case for nodes migrating to using rocksdb, maybe the message in the panic is not clear enough, or we want to let them unwind to 0 given that they will need to resync in any case, correct?

also, i see the comment "Highly unlikely to happen, and given its destructive nature, it's better to panic" given that it is more likely to happen now, if we want this behavior we should update the comment, wdyt?

yongkangc committed:
fix: allow RocksDB unwind to block 0 for migration scenario

Address review feedback: nodes migrating to RocksDB will commonly trigger
unwind to block 0, which is expected. Only panic if static files also
require unwind to 0 (indicating corruption). Otherwise, warn and proceed
with full resync.

yongkangc committed:
fix: address Joshi's review comments

- Revert 'allow RocksDB unwind to 0' - keep panic on unwind to 0 as it would
  leave MDBX with huge free list size
- Remove checked_add overflow protection (not needed in practice)
- Created issue #20506 for refactoring StaticFileProvider::check_consistency

Amp-Thread-ID: https://ampcode.com/threads/T-019b3076-f7ed-700d-8916-779712facb8e

yongkangc commented on crates/storage/provider/src/providers/rocksdb/invariants.rs:
```diff
@@ -120,7 +127,12 @@ impl RocksDBProvider {
                 );
-                self.prune_transaction_hash_numbers_in_range(provider, (mdbx_tx + 1)..=highest_tx)?;
+                // Use checked_add to prevent overflow when mdbx_tx == u64::MAX
+                if let Some(start) = mdbx_tx.checked_add(1) {
```
i guess we wont ever reach there haha

addressed here [2af821d](https://github.com/paradigmxyz/reth/pull/20478/commits/2af821d5fb30fd5736b0367d433e6b46aabadab0)

yongkangc committed:
fix: update consistency check ordering per Joshi's feedback

Reorder consistency checks as follows:
1. StaticFileProvider::check_file_consistency() - heals interrupted writes
2. RocksDB::check_consistency() - uses static file data for tx hash pruning
3. StaticFileProvider::check_consistency() - compares with MDBX checkpoints

This ensures RocksDB can access transaction data before static files prune it.

Amp-Thread-ID: https://ampcode.com/threads/T-019b3076-f7ed-700d-8916-779712facb8e

yongkangc committed:
fix: make stub iter return empty iterator instead of error

This is consistent with first() and last() returning Ok(None) -
the stub behaves as if the database is empty rather than unavailable.

yongkangc committed:
fix: adapt to #20508 API and add const to stub functions

- Remove non-existent check_file_consistency() call
- Simplify ordering: static file check first, then RocksDB check
- Add const to first() and last() stub functions

yongkangc committed:
del

yongkangc committed:
feat(provider): add check_file_consistency for proper ordering

Addresses Joshi's review comment about consistency check ordering.

The ordering is now:
1. check_file_consistency() - heals NippyJar files without pruning
2. rocksdb check_consistency() - needs static file tx data
3. static file check_consistency() - compares with MDBX, may prune

This ensures RocksDB can access transaction data before static files
potentially prune it during checkpoint comparison.

yongkangc committed:
fix: unwind when rocksdb history lags checkpoint

yongkangc committed:
reverted

yongkangc committed:
comment

yongkangc commented on crates/node/builder/src/launch/common.rs:
```diff
@@ -501,20 +502,61 @@ where
             );
 
-            info!(target: "reth::cli", unwind_target = %unwind_target, "Executing an unwind after a failed storage consistency check.");
+            let unwind_target = PipelineTarget::Unwind(unwind_block);
```
we need this now as we are combining 3 different targets
